### PR TITLE
Allow property bags in TimeZone.from and Calendar.from

### PR DESCRIPTION
**Pull request by ptomato**
Closes: #925

### CODE REVIEW

File: polyfill/lib/calendar.mjs [hunk 1]
```diff
@@ -117,7 +117,11 @@ export class Calendar {
     return GetSlot(this, CALENDAR_ID);
   }
   static from(item) {
-    if (ES.Type(item) === 'Object') return item;
+    if (ES.Type(item) === 'Object') {
+      if (!('calendar' in item)) return item;
```
Comment (Ms2ger) [inline polyfill/lib/calendar.mjs:121]: Prefer getting the property once and checking against `undefined`.

File: polyfill/lib/calendar.mjs [hunk 2]
```diff
@@ -117,7 +117,11 @@ export class Calendar {
...
+      item = item.calendar;
+      if (ES.Type(item) === 'Object' && !('calendar' in item)) return item;
```
Comment (Ms2ger) [inline polyfill/lib/calendar.mjs:123]: What's the use case for a calendar property that's not a calendar object yet?

File: polyfill/lib/timezone.mjs [hunk 3]
```diff
@@ -236,7 +236,11 @@ export class TimeZone {
     return ES.TimeZoneToString(this);
   }
   static from(item) {
-    if (ES.Type(item) === 'Object') return item;
+    if (ES.Type(item) === 'Object') {
+      if (!('timeZone' in item)) return item;
+      item = item.timeZone;
+      if (ES.Type(item) === 'Object' && !('timeZone' in item)) return item;
```
Comment (Ms2ger) [inline polyfill/lib/timezone.mjs:242]: Ditto

**Review by Ms2ger — APPROVED** [review]
At this point, I'm wondering if we should just require subclassing `Temporal.Calendar`/`Temporal.TimeZone` for custom implementations. What's the value of allowing plain objects, and does it justify this?

**Comment by ptomato** [comment]
> At this point, I'm wondering if we should just require subclassing `Temporal.Calendar`/`Temporal.TimeZone` for custom implementations. What's the value of allowing plain objects, and does it justify this?

I had this same question, but Richard answered it in the last Temporal meeting. Calendars and TimeZones from other realms will be considered plain objects, so that's the motivation for this functionality.

File: polyfill/lib/calendar.mjs (hunk 1 above)
Comment (ptomato) [reply polyfill/lib/calendar.mjs:121]: I think that considering an object with `calendar: undefined` to signify a custom calendar would work, but it is different from the semantic that we landed on, where it's the shape of the object that matters: if a `calendar` property is present then it's a property bag, if absent then it's a custom calendar.

File: polyfill/lib/calendar.mjs (hunk 2 above)
Comment (ptomato) [reply polyfill/lib/calendar.mjs:123]: The calendar could be given as a string in the property bag:
```js
const myPropertyBag = { year: 2020, month: 10, day: 20, calendar: 'gregory' };
Temporal.Calendar.from(myPropertyBag);
```

**Comment by ljharb** [comment]
Not if they inherit from a base TimeZone class that has an internal slot.

**Comment by ptomato** [comment]
As far as I understand, Richard felt that a brand check was not correct to use in this case - @gibson042 can you clarify?

**Comment by ljharb** [comment]
Subclassing (with one weird exception in RegExp) generally always uses a brand check. The `Set.prototype.size` getter, for example, can be `.call`ed on Set instances from other realms, because it checks the internal slot. Similarly, `[arr].flat()` where `arr` is an array with a single item, will result in an array with that single item, as long as `arr` passes `IsArray` in the spec, which detects cross-realm arrays as well.

**Review by Ms2ger — APPROVED** [review]
I think we should reconsider plain objects, but that's for another bug.

**Pull request merged**